### PR TITLE
Resource resolvers

### PR DESCRIPTION
**Pull request by cd1989**
**What this PR does / why we need it**:

Implements basic resource resolvers.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @zhujian7 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: build/workflow/resolver/image/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/sh
+set -e
+
+USAGE=$(cat <<-END
+    This tool is used to resolver image resources, here image resource
+    stands for an image in docker registry.
+
+    Usage:
+        $ docker run -it --rm \\
+            -e IMAGE=test.caicloudprivatetest.com/library/alpine:3.6 \\
```
Comment (ddysher) [inline build/workflow/resolver/image/entrypoint.sh]: use a public image

File: build/workflow/resolver/git/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
+FROM alpine/git:1.0.4
+
+ENV WORKDIR /workspace
+WORKDIR $WORKDIR
+
+RUN mkdir -p /workspace/data
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/git/Dockerfile]: OK, I will make it configurable and keep this as default.

File: build/workflow/resolver/kv/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+if [[ $# != 1 ]]; then
+    echo "$USAGE"
+    exit 1
+fi
+COMMAND=$1
+
+# Check whether required environment variables are set.
+if [ -z ${CYCLONE_ADMIN_URL+x} ]; then echo "CYCLONE_ADMIN_URL is unset"; exit 1; fi
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/kv/entrypoint.sh]: you are right, we can have the default value.

File: build/workflow/resolver/kv/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+if [ -z ${WORKFLOWRUN+x} ]; then echo "WORKFLOWRUN is unset"; exit 1; fi
+if [ -z ${STAGE+x} ]; then echo "STAGE is unset"; exit 1; fi
+
+pull() {
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/kv/entrypoint.sh]: Embarrassed

File: build/workflow/resolver/kv/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    if [ ! -e "/workspace/data/kv.txt" ]; then
+        echo "Key-value file /workspace/data/kv.txt not exist."
+        exit 1
+    fi
+
+    # Convert key-values to json.
+    items=$(sed -n '/\S/p' /workspace/data/kv.txt | sed -r 's/^(.+):\s*(.+)$/{"key":"\1","value":"\2"}/g')
+    json=$(printf '{"items":[%s]}' $(echo ${items} | sed -r 's/\}\s*\{/\},\{/g'))
+
+    echo "[PUT] ${CYLONE_ADMIN_URL}/workspacerun/${WORKFLOWRUN}/stage/${STAGE}/kv"
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/kv/entrypoint.sh]: Name of a WorkflowRun

File: build/workflow/resolver/kv/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+set -e
+
+USAGE=$(cat <<-END
+    This tool is used to resolve key-value resources, here kv resource
+    stands for a set of key-values.
+
+    Usage:
+        $ docker run -it --rm \\
+            -e KV_PATH=/workspace/data/kv.txt \\
+            -e CYCLONE_ADMIN_URL=cyclone-admin \\
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/kv/entrypoint.sh]: ok

File: build/workflow/resolver/kv/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+fi
+COMMAND=$1
+
+# Check whether environment variables are set.
+if [ -z ${CYCLONE_ADMIN_URL+x} ]; then echo "WARN: CYCLONE_ADMIN_URL is unset, use default 'cyclone-admin'"; fi
+if [ -z ${WORKFLOWRUN+x} ]; then echo "WORKFLOWRUN is unset"; exit 1; fi
+if [ -z ${STAGE+x} ]; then echo "STAGE is unset"; exit 1; fi
+
+push() {
+    if [ ! -e "/workspace/data/kv.txt" ]; then
```
[the remark this answers is not in the document]
Comment (cd1989) [reply build/workflow/resolver/kv/entrypoint.sh]: My mistake

**Pull request merged**